posts/2018-08-16-well-typed-heap.lhs

```diff
@@ -0,0 +1,1100 @@
+maybe a bit naughty but this prevents us from carrying around the singletons in
+a bunch of places.
+
+> treesToSBin :: Forest o b a -> SBin b
+> treesToSBin FEnd     = SBEnd
+> treesToSBin (F0 t)   = SB0 (treesToSBin t)
+> treesToSBin (F1 _ t) = SB1 (treesToSBin t)
+
+popHeap: gluing the pieces together
+-----------------------------------
+
+We can now find all trees in the heap that may be cut.  They are returned in a
+`CutTree` datatype.  If we assume that we are taking a specific `CutTree`, we
+can take the root from the tree inside this datatype, and we can construct a new
+heap from its children using `childrenToForest`.  Then, we merge it back
+together with the original heap.
+
+The new heap has one less element -- hence we use `BDec` (binary decrement,
+define just a bit below).
+
+> popForest
+>     :: forall a b. Ord a
+>     => CutTree 'Zero b a
+>     -> (a, Forest 'Zero (BDec b) a)
+
+We deconstruct the `CutTree` to get the root (`x`) of the selected tree,
+the children of the selected trees (`children`), and the remaining trees in the
+heap (`forest`).
+
+> popForest (CutTree
+>             xnat _nnat
+>             (Tree x (children :: Children r a))
+>             (forest :: Forest 'Zero l a)) =
+
+We construct a new forest from the children.
+
+>     let cforest = childrenToForest xnat children
+
+We merge it with the remainder of the heap:
+
+>         merged :: Forest 'Zero (BAdd l (Ones r)) a
+>         merged = mergeForests forest cforest
+
+The illustration from above applies here:
+
+![Merging back together](/images/draft-08.png)
+
+Now, we cast it to the result using a new `lemma4` with a singleton that we
+construct from the trees:
+
+>         evidence :: SBin (BAdd l (Ones r))
+>         evidence = treesToSBin merged in
+>     (x, case lemma4 evidence of QED -> merged)
+
+This is the type family for binary decrement.  It is partial, as expected -- you
+cannot decrement zero.  This is a bit unfortunate but necessary.  Having the
+`BNonZero` type family and using it as a constraint will solve that though.
+
+> type family BDec (binary :: Binary) :: Binary where
+>     BDec ('B1 b) = 'B0 b
+>     BDec ('B0 b) = 'B1 (BDec b)
+
+The weirdly specific `lemma4` helps us prove that we can take a number,
+increment it and then decrement it, and then get the same number back _provided_
+incrmenting doesn't change its width.  This ends up matching perfectly with the
+width constraint generated by the `CutTree`, where the number that we increment
+is a number of "1"s smaller than the shape of the total heap (intuitively).
+
+Using another constraint in `CutTree` with another proof here should also
+be possible.  I found this it hard to reason about _why_ this constraint is
+necessary, but once I understood that it wasn't too abnormal.  The proof is easy
+though.
+
+> lemma4
+>     :: (Width x ~ Width (BInc x))
+>     => SBin x
+>     -> BDec (BInc x) :~: x
+> lemma4 (SB0 _) = QED
+> lemma4 (SB1 b) = case lemma4 b of QED -> QED
```

Comment:
You don't get a warning about the unmatched `SBEnd` here?

Reply:
Added a note